New function added for reading  an entire file

Sayan-Chaudhuri opened the pull request:
The new function is only concerned with reading an entire file , previously handled within the qio_channel_read_string function. The function also handles cases where file size is not known beforehand . The reading performance has been improved drastically to match the likes of C++ .

e-kayrakli committed:
Drop unnecessary calls to c_str in compiler

Signed-off-by: Engin Kayraklioglu <e-kayrakli@users.noreply.github.com>

Sayan-Chaudhuri committed:
New function added for reading  an entire file

The new function is only concerned with reading an entire file , previously handled within the qio_channel_read_string function. The function also handles cases where file size is not known beforehand . The reading performance has been improved drastically to match the likes of C++ .

Requested changes made in read_entire_file function

comments added for requested changes

Changes Made as Requested

Changes Made as Requested

Added capability to read the file even if it's size grows while reading

Added comments for better readability

Made the requested changes

New bool type variable introduced

set err equal to QIO_EEOF

updated chunkSize, while loop reorientation,file_length limited to initial section

Limited code to 80 columns

Line 402 fit to 80 columns

Comments clean up in qio_channel_read_string function

Removed null check on ret

Doubled chunkSize to prevent reallocation complexity becoming quadratic

Comment indented [386-389], added break on line 512

Sayan-Chaudhuri commented:
@mppf

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -417,14 +484,14 @@ qioerr qio_channel_read_string(const int threadsafe, const int byteorder, const
-      // Ignore EOF errors as long as we read something.
-      if( err && qio_err_to_int(err) == EEOF && num > 0 ) err = 0;
-      break;
+//     case QIO_BINARY_STRING_STYLE_TOEOF:
```
Probably best to leave the case here but have it assert and break?

Sayan-Chaudhuri replied:
What should I assert for ?

mppf replied:
If you are asking "Why bother have it assert here?" it's just to make it easier to maintain. The case should be handled above and if it isn't, there is a bug.

Sayan-Chaudhuri replied:
I apologise Sir but if you can kindly explain the point a bit more ? I mean what is it that I should be checking in assert ? The assertion will be that this case has already been handled ?? but then , in that case,it won't be entering the if statement because I am returning from that if block itself .

mppf replied:
Yes, the case should be handled above. If it is not, there is a bug, and that will be easier to fix if there is an assert. Think of it as a way to help future developers making changes in this area.

Sayan-Chaudhuri replied:
Ok, I shall make this change once you are able to see the other changes I have made .

mppf replied:
Still need to do something here

Sayan-Chaudhuri replied:
@mppf  I have added a break statement . Shouldnt that suffice ?

mppf replied:
it is fine. next step is to merge all of your commits into one and make a PR against the main Chapel repository.

Sayan-Chaudhuri replied:
@mppf  Ok..Kindly allow me some days for this.

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -360,6 +360,68 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
 }
+qioerr read_entire_file(qio_channel_t* restrict ch, const char* restrict* restrict out, int64_t* restrict len_out, ssize_t maxlen)
+{
+    qioerr err,err_reading_file;
```
indenting in this function is a bit weird. Please use two spaces per indent level and make sure each section in curly braces has a new indent level.

Sayan-Chaudhuri replied:
Changes made as requested .

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -395,7 +457,12 @@ qioerr qio_channel_read_string(const int threadsafe, const int byteorder, const
+    {
+       //printf("Inside the new if condition");
+       err=read_entire_file(ch,out,len_out,maxlen);
+       return err;
```
Returning here will cause the channel never to be unlocked.
Consider one of these two options:
 1. Use `goto unlock` here and make sure the following logic makes sense in this setting or
 2. Move your check for `QIO_BINARY_STRING_STYLE_TOEOF` and the call to `read_entire_file` before the `if( threadsafe )` block above that does the locking, and then handle locking and unlocking inside of `read_entire_file`.

I think I have a slight preference for 2.

Either way, your new function should be called before the `qio_channel_mark` call.

Sayan-Chaudhuri replied:
Changes made as requested .

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -360,6 +360,68 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+        
+        int64_t fileLenGuess=1024;
+        ret=qio_malloc(fileLenGuess+1);
+        while(true)
```
I'd like it to work correctly if the file size is wrong or if it changes during the read. To do that, I'd like to see a while loop like this after the read we have above.

A code outline like this:
 * get file size
 * if that failed, choose a guess for initial buffer size
 * allocate initial buffer
 * loop reading in to buffer and reallocating

Sayan-Chaudhuri replied:
Made the necessary changes

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -360,6 +360,68 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+          
+
+    err_reading_file = 0;
+    rewind:
```
this rewind block should not be needed here because this function should be run before mark is called in the calling function

Sayan-Chaudhuri replied:
Changes made as requested ,

mppf replied:
The rewind block is still there. Have you tried looking at https://github.com/Sayan-Chaudhuri/chapel/pull/1/files yourself?

Sayan-Chaudhuri replied:
@mppf  Sir, request you to kindly have a look at this link - https://github.com/Sayan-Chaudhuri/chapel/pull/1/commits/952cf98a6c9e9c8fef27a8f501cfea140eab98c0.  I had already made the corrections and linked below with the name "comments added for requested changes"  .

mppf replied:
Yes, I see no rewind block when I look at that commit, but https://github.com/Sayan-Chaudhuri/chapel/pull/1/files still shows it. Can you try to figure out why?

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -360,6 +360,68 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+        {
+            ret[0]='\0';
+            err_reading_file = qio_channel_read(false, ch, ret, fileLenGuess, &amt);
+            if (err_reading_file == QIO_EEOF) break;
```
If we have reached EOF, amt will be some final amount we read. In that case, we should probably reallocate the buffer to only include this much rather than zeroing everything else. Make sure to leave space for the trailing zero byte.

Sayan-Chaudhuri replied:
Made the changes .

Sayan-Chaudhuri commented:
@mppf  I shall try to be ready with the changes you proposed by this Thursday .

Sayan-Chaudhuri commented:
@mppf  Request you to kindly have a look at the requested changes .

mppf commented:
@Sayan-Chaudhuri - I don't know what the problem is but it doesn't look like any of the requested changes appear in the version showing in this PR.

Sayan-Chaudhuri commented:
@mppf Kindly let me know if you can view the changes now

Sayan-Chaudhuri commented:
Request you to kindly check the file "comments added for requested changes"

Sayan-Chaudhuri commented:
@mppf Please see the link just below this comment "Changes Made as Requested". Editing in the link you provided,  always lead me to create a new commit with a new name .

Sayan-Chaudhuri commented:
@mppf  Were you able to view the changes I made ? I would love to know if there's any scope for improvement

mppf commented:
hi @Sayan-Chaudhuri when I'm reviewing a code change / PR,  it's important to me to see the change that is proposed in the Files view where I will look at the diff. This is important because I need to be able to see all of the changes in different commits together. In this PR I am just seeing that the qio_formatted.c file is deleted. If you are running into challenges using git / github you can read about these online. We have some documentation on these within the Chapel project on https://chapel-lang.org/docs/developer/bestPractices/git.html but there are many other online resources that you might find helpful.

Sayan-Chaudhuri commented:
@mppf  Ok, I got it now...Thanks for the resources

Sayan-Chaudhuri commented:
@mppf  fixed the issue of deleted file. 
These are the changes made from the first commit -
* uniform indenting with 2 spaces per indent level
* handle locking and unlocking inside read_entire_file
* reallocation of buffer if file length cannot be obtained or addition to file contents while reading .
* removed rewind block
* removing unused buffer space .

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,86 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+  }
+
+  err = qio_channel_mark(false, ch);
+  if( err ) goto unlock;
```
You shouldn't have these two lines, since we don't need to mark, and we removed the code to revert/commit.

Sayan-Chaudhuri replied:
@mppf  I would like to further the discussion on locking here . Initially, we had discussed that the locking and unlocking of channel to be handled inside the read_entire_file function . My question here is, why do we need to lock the channel at all ? We are writing code to handle read/write concurrently at the same time . But locking will not allow that ,read/write on the same resource simultaneously .

mppf replied:
the locking is to protect the channel data structure

Sayan-Chaudhuri replied:
OK

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,86 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+    ret=qio_malloc(file_length+1);
+    ret[0]='\0';
+    err=qio_channel_read(false, ch, ret, file_length, &amt);
+  }
```
If the file length grew while we did this read, the err will be 0 but we still need to loop.

Sayan-Chaudhuri replied:
But then  what can be the terminating condition ? Because, there is the chance that each time we are trying to read the file, its length is getting increased .

Sayan-Chaudhuri replied:
@mppf

Sayan-Chaudhuri replied:
@mppf  Also,why this happens  'If the file length grew while we did this read, the err will be 0  ' ?

mppf replied:
if the read reached EOF then qio_channel_read will return EEOF and returns a short amount read (maybe 0 or maybe more).

This is in a documentation comment on that function
`// On a read, returns EEOF and *amt_read=some amount for EOF.`

Sayan-Chaudhuri replied:
OK.Thanx

Sayan-Chaudhuri replied:
@mppf  I checked and I found that qio_channel_read is not returning EEOF if the file read is completed . I am not sure if I am making any mistake. I checked in the original code for the qio_channel_read_string  function whether the value of err in line 463 becomes equal to EEOF after the read is completed but this condition (qio_err_to_int(err)==EEOF) is becoming false there.

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,86 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+    while(true)
+    {
+      ret[0]='\0';
+      err_reading_file = qio_channel_read(false, ch, ret, fileLenGuess, &amt);
```
you need to track how much you have already read (which each call will store into amt) and read into `&ret[pos]`. You can't just write `&ret[amt]` because amt is just the amount read from this read, but you need the total from all reads.

mppf replied:
As it is now, you will be replacing the data in the first portion of `ret` but not filling in the rest

Sayan-Chaudhuri replied:
@mppf  Sincere apologies but I am not exactly getting the idea here . Let's say, with the initial buffer size we are able to read the first line of the file. But the file has 2 lines. So, then , our program will double the buffer capacity and on the next read, read the first line again and along with it read the second line as well.

Sayan-Chaudhuri replied:
Sorry Sir, I have got my mistake ..

mppf commented:
> @mppf I checked and I found that qio_channel_read is not returning EEOF if the file read is completed . I am not sure if I am making any mistake. I checked in the original code for the qio_channel_read_string function whether the value of err in line 463 becomes equal to EEOF after the read is completed but this condition (qio_err_to_int(err)==EEOF) is becoming false there.

I got this in an email but can't find the comment. Anyway if you read some data then it won't return EOF. It should only return EOF when there is no data read.

Sayan-Chaudhuri commented:
@mppf  I think I have understood it finally . Thanks

Sayan-Chaudhuri commented:
@mppf Made the changes as suggested
1) Added capability to continue file reading if it grows in size even if the error is zero
2) Added code to track the total amount of read happening over several iterations in case of file size growing in runtime.

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,88 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+    }
+    else
+    {
+       err=qio_int_to_err(1);
```
seems like you should use a different variable of `int` or `bool` type rather than trying to store 1 as an error

Sayan-Chaudhuri replied:
@mppf I changed it from 1 to true . Will it be okay ? Also,like what can be the potential issue if we keep 1 there ?

mppf replied:
1 is not an OS error code which is what err_t is supposed to store

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,88 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+  ret[file_length] = '\0'; // always add terminator at the end
+    
+  err= 0;
+  unlock:
```
let's remove the `unlock:` since it appears unused

Sayan-Chaudhuri replied:
removed

mppf reviewed (commented):
At this point mainly it needs some help with comments. There is a lot of commented out code making it harder to follow and comments describing why it is taking the various steps are missing.

Sayan-Chaudhuri commented:
Ok. I am commenting it just now.

Sayan-Chaudhuri commented:
@mppf  Made the changes
1) Added comments .
2) deleted unlock block .
3) Changed the constant from 1 to 'true'

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,83 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+    }
+    else
+    {
+       err=qio_int_to_err(true);
```
I was saying you should introduce another variable, besides `err`, to do this; and that the new variable should have type `int` or `bool`.

Sayan-Chaudhuri commented:
@mppf  I have now used a error flag variable 'err_flag'  of boolean type instead of using any constants .

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,85 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+    else
+    {
+       err_flag=true;
+       err=qio_int_to_err(err_flag);
```
this line is still wrong since `1` is not an OS error code which is what `err_t` stores

Sayan-Chaudhuri replied:
@mppf  Ok,I think I need to understand this setting of error a bit more clearly.

Sayan-Chaudhuri replied:
@mppf I have set it to QIO_EEOF. I believe that should do now. It was a silly mistake from my side .

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,84 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+   
+    ret=qio_malloc(file_length+1);              //allocate initial buffer
+    ret[0]='\0';
+    err=qio_channel_read(false, ch, ret,file_length+1, &amt);    //file_length+1 instead of file_length -->I try to read one more character than the prescribed 
```
We keep the source code to 80 column lines. Could you move the comments to be before the lines that they apply to?

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,84 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+                                                                 // character.
+    if(qio_err_to_int(err)==EEOF)                                // Here,checking whether file end is reached.
+    {
+       err=0;
```
If the file became shorter between the qio_file_length and the qio_channel_read call then we will have allocated more than necessary and could realloc here.

mppf replied:
(I am not sure that case matters, though.)

Sayan-Chaudhuri replied:
@mppf In my opinion, I do not feel this case is valid. Through programs, we can read and write a file, deletion of file contents does not happen in the truest sense, we basically overwrite file contents, so the entire contents of the previous version becomes invalid. So,that way,if we speak of file shortening, we should be rereading the entire file instead of reallocating to prevent wastage of memory.

mppf replied:
I don't really believe that argument but I don't think you need to handle the case because allocating extra memory in that rare case seems like an issue we don't have to solve here.

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,84 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+    ret=qio_realloc(ret,total_amt_read+fileLenGuess+1);            // add some more space to the buffer to read the remaining data.
+    while(true)
+    {
+      err_reading_file = qio_channel_read(false, ch,&ret[total_amt_read],1024LL, &amt);   // we will read the remaining data in units of 1024 each time .
```
Can we keep 1024 in a variable called, say, `chunkSize` ? Would like to only change it in one place if we need to change it.

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,84 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+      total_amt_read+=amt;                                                                //calculate the amt that we have read until now
+      if (err_reading_file == QIO_EEOF) break;                                            // break out of loop is data is exhausted finally
+      fileLenGuess=total_amt_read+1024;
+      ret = qio_realloc(ret, fileLenGuess+1);
```
IMO the code will be clearer if you put these last two lines at the top of the body of the while(true) loop and remove the realloc before the loop.

Sayan-Chaudhuri replied:
@mppf  I am sorry but I did not get the idea here. Should I make a copy of lines 410 and 411 and place that before the while loop ? lines 410 and 411 cannot be removed completely because then we won't be able to read a dynamically changing file size

mppf replied:
Right I was just asking you to write the loop differently.

IMO you have

```
  DoThingA();
  while (true) {
    DoThingB();
    DoThingA();
  }
```

and I am asking you to change it to


```
  while (true) {
    DoThingA();
    DoThingB();
  }
```

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,84 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+      fileLenGuess=total_amt_read+1024;
+      ret = qio_realloc(ret, fileLenGuess+1);
+    }
+    file_length=total_amt_read;                                                            //get the new file length
```
I think the code will be clearer if, in this line and the rest of the function, you just use `total_amt_read` instead of the `file_length` variable. (So, remove line 413, use total_amt_read in 414 and 419). That way, file_length is only relevant for the first part of the function.

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,84 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+
+
+  errcode = qio_err_to_int(err);
+  if( errcode && errcode != EEOF && errcode != ESHORT ) qio_free(ret);
```
I don't think ESHORT is possible to get here so you can remove that case

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -387,15 +465,21 @@ qioerr qio_channel_read_string(const int threadsafe, const int byteorder, const
+  
+  if(str_style==QIO_BINARY_STRING_STYLE_TOEOF)  //if block moved before threadsafe
+  {
+    //printf("Inside the new if condition");
```
Looks like there are comments to clean up in this function too

mppf reviewed (commented):
@Sayan-Chaudhuri - thanks for working through these changes. If you are able to address these comments then you can make a PR against the main Chapel repo with these changes. One thing you will need to do is to add a Chapel test of this code. You can read about how to do that here: https://chapel-lang.org/docs/developer/bestPractices/TestSystem.html#readme-testsystem

Sayan-Chaudhuri commented:
> @Sayan-Chaudhuri - thanks for working through these changes. If you are able to address these comments then you can make a PR against the main Chapel repo with these changes. One thing you will need to do is to add a Chapel test of this code. You can read about how to do that here: https://chapel-lang.org/docs/developer/bestPractices/TestSystem.html#readme-testsystem

@mppf  Thanks for reviewing my work and giving valuable inputs time to time. I shall make the necessary changes and get back on Monday

Sayan-Chaudhuri commented:
@mppf Made the changes as suggested.

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -364,31 +364,32 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+    //file_length+1 instead of file_length -->I try to read one more character than the prescribed 
+   //file length . That way, if we have already reached the file end, reading 1 more character will 
+   //will generate EOF. But, if the file length has increased while reading, we have just read one more
+   // character.
```
Something weird happened with the spacing. 

Also as a style matter we don't normally put `I` in the comments (because nobody knows who that is when reading the comment). So you can say e.g.

"Why file_length+1 instead of file_length? Read one more character than the prescribed 
file length. That way, ..."

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,91 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+      err_reading_file = qio_channel_read(false, ch,&ret[total_amt_read],chunkSize, &amt);   
+      total_amt_read+=amt;                                                                
+      if (err_reading_file == QIO_EEOF) break;                                            
+      
```
This loop used to double chunkSize on each iteration, you could do that here at the end simply with `chunkSize *= 2`. That prevents the reallocation process from becoming O(n*n).

Sayan-Chaudhuri replied:
@mppf What I am basically doing is reading 1024 characters each time. So,why would the relocation be O(n*n)?

mppf replied:
the call to realloc is O(size of memory being reallocated). You are calling it in a loop that is O(size of memory / 1024). So it is O(size * size).

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,91 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+  else 
+  {
+    // don't modify out if we didn't read anything.
+    if( ret ) 
```
it is not possible for `ret` to be NULL at this point

mppf reviewed (commented):
You have to make this a pull request at https://github.com/chapel-lang/chapel/pulls . As you do that, can you handle my feedback and then collapse the commit history? (You can do that either manually by copying this file to a new checkout -- or with `git rebase -i`. You can find resources about how to do that online if you are not familiar).

Sayan-Chaudhuri commented:
@mppf I have made the necessary changes to chunkSize as you said. Please kindly let me know of any further corrections needed ,otherwise I would like to proceed for converting this into a pull request for the main repo.

mppf commented on runtime/src/qio/qio_formatted.c:
```diff
@@ -361,6 +361,89 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
+   //Why file_length+1 instead of file_length? When the program tries to read one more character than 
+   //the prescribed file length then, if we have already reached the file end,
+   //reading 1 more character will generate EOF. But, if the file length has increased
+   //while reading, we have just read one more character.
```
please indent this comment to match above code

Sayan-Chaudhuri commented:
@mppf   I have indented the comments.
